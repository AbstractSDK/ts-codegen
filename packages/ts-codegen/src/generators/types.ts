import { pascal } from "case";
import { header } from '../utils/header';
import { join } from "path";
import { sync as mkdirp } from "mkdirp";
import * as t from '@babel/types';
import { writeFileSync } from 'fs';
import generate from "@babel/generator";
import { clean } from "../utils/clean";
import { findAndParseTypes, findExecuteMsg, getDefinitionSchema } from '../utils';
import { cosmjsAminoImportStatements } from '../utils/imports';
import { RenderContext, TSTypesOptions } from "wasm-ast-types";
import { BuilderFile } from "../builder";

export default async (
  name: string,
  schemas: any[],
  outPath: string,
  tsTypesOptions?: TSTypesOptions
): Promise<BuilderFile[]> => {

  const context = new RenderContext(getDefinitionSchema(schemas), {
    tsClient: tsTypesOptions ?? {}
  });
  const options = context.options.types;

  const localname = pascal(name) + '.types.ts';
  const ExecuteMsg = findExecuteMsg(schemas);
  const typeHash = await findAndParseTypes(schemas);

  const body = [];

  body.push(cosmjsAminoImportStatements(typeHash));

  // TYPES
  Object.values(typeHash).forEach((type: t.Node) => {
    body.push(
      clean(type)
    )
  });

  // alias the ExecuteMsg
  if (options.aliasExecuteMsg && ExecuteMsg) {
    body.push(
      t.exportNamedDeclaration(
        t.tsTypeAliasDeclaration(
          t.identifier(`${name}ExecuteMsg`),
          null,
          t.tsTypeReference(t.identifier('ExecuteMsg'))
        )
      )
    );
  }

  const code = header + generate(
    t.program(body)
  ).code;

  mkdirp(outPath);
  const filename = join(outPath, localname);
  writeFileSync(filename, code);

  return [
    {
      type: 'type',
      contract: name,
      localname,
      filename,
    }
  ]
};
