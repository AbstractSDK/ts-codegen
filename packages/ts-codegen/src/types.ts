import { TSClientOptions, ReactQueryOptions } from "wasm-ast-types";

export interface TSBuilderOptions {
    contractDirs: string[];
    outPath: string;

    // TODO: separate types
    typesOnly?: boolean;

    reactQuery?: ReactQueryOptions & { enabled: true };
    tsClient?: TSClientOptions & { enabled: true };
};