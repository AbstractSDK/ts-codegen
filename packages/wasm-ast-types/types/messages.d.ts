import * as t from '@babel/types';
import { ExecuteMsg } from './types';
import { RenderContext } from './utils/types';
export declare const createFromPartialClass: (context: RenderContext, className: string, implementsClassName: string, execMsg: ExecuteMsg) => t.ExportNamedDeclaration;
export declare const createFromPartialInterface: (context: RenderContext, className: string, execMsg: ExecuteMsg) => t.ExportNamedDeclaration;
