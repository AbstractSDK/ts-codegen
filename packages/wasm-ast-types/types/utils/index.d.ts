export * from './babel';
export * from './types';
export * from './ref';
