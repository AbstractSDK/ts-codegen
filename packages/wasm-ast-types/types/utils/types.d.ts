import * as t from '@babel/types';
import { TSTypeAnnotation } from '@babel/types';
export interface RenderOptions {
}
export interface RenderContext {
    schema: JSONSchema;
    options: RenderOptions;
}
export declare class RenderContext implements RenderContext {
    schema: JSONSchema;
    constructor(schema: JSONSchema, options?: RenderOptions);
    refLookup($ref: string): JSONSchema;
}
export interface JSONSchema {
    $ref?: string;
    $schema?: string;
    additionalProperties?: boolean;
    allOf?: JSONSchema[];
    anyOf?: JSONSchema[];
    definitions?: Record<string, JSONSchema>;
    description?: string;
    oneOf?: JSONSchema[];
    properties?: Record<string, JSONSchema>;
    patternProperties?: Record<string, JSONSchema>;
    items?: JSONSchema[] | JSONSchema;
    additionalItems?: JSONSchema;
    required?: string[];
    title?: string;
    type?: string;
}
export declare const getTypeFromRef: ($ref: any) => t.TSTypeReference;
export declare const getType: (type: any) => t.TSBooleanKeyword | t.TSNumberKeyword | t.TSStringKeyword;
export declare const getPropertyType: (context: RenderContext, schema: JSONSchema, prop: string) => {
    type: any;
    optional: boolean;
};
export declare function getPropertySignatureFromProp(context: RenderContext, jsonschema: JSONSchema, prop: string, camelize: boolean): {
    type: string;
    key: t.Identifier;
    typeAnnotation: t.TSTypeAnnotation;
    optional: boolean;
};
export declare const getParamsTypeAnnotation: (context: RenderContext, jsonschema: any, camelize?: boolean) => t.TSTypeAnnotation;
export declare const createTypedObjectParams: (context: RenderContext, jsonschema: JSONSchema, camelize?: boolean) => t.ObjectPattern;
