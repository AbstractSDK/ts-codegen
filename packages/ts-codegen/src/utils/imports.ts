import * as w from 'wasm-ast-types';

export const cosmjsAminoImportStatements = (typeHash) => {
  const cosmjsImports = ['StdFee'];

  // If the coin is not provided by the generation already, import it from cosmjs
  if (!typeHash.hasOwnProperty('Coin')) {
    cosmjsImports.unshift('Coin');
  }

  return w.importStmt(cosmjsImports, '@cosmjs/amino');
};
