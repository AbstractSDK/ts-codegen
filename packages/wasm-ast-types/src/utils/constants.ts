import { identifier } from './babel';
import * as t from '@babel/types';

export const OPTIONAL_FUNDS_PARAM = identifier(
  'funds_',
  t.tsTypeAnnotation(t.tsArrayType(t.tsTypeReference(t.identifier('Coin')))),
  true
);

export const FEE_PARAM = identifier(
  'fee_',
  t.tsTypeAnnotation(
    t.tsUnionType([
      t.tsNumberKeyword(),
      t.tsTypeReference(t.identifier('StdFee')),
      t.tsLiteralType(t.stringLiteral('auto'))
    ])
  )
);

export const OPTIONAL_FEE_PARAM: t.Identifier = {
  ...FEE_PARAM,
  optional: true
};

export const OPTIONAL_MEMO_PARAM = identifier(
  'memo_',
  t.tsTypeAnnotation(t.tsStringKeyword()),
  true
);

export const OPTIONAL_FIXED_EXECUTE_PARAMS = [
  OPTIONAL_FEE_PARAM,
  OPTIONAL_MEMO_PARAM,
  OPTIONAL_FUNDS_PARAM
];

export const PROVIDER_TYPES = {
  SIGNING_CLIENT_TYPE: "client",
  QUERY_CLIENT_TYPE: "queryClient",
  MESSAGE_COMPOSER_TYPE : 'message-composer',
  PROVIDER_TYPE : 'provider'
};
