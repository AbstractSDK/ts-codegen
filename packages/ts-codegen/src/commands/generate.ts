import { prompt } from '../utils/prompt';
import codegen from '../index';
import { TSBuilderOptions } from '../builder';

export default async (argv) => {
    const questions = [
        {
            _: true,
            type: 'path',
            name: 'schema',
            message: 'which directory contains the the Rust contracts?',
            default: './schema'
        },
        {
            _: true,
            type: 'path',
            name: 'out',
            message: 'where is the output directory?',
            default: './ts'
        },
        {
            _: true,
            type: 'string',
            name: 'name',
            message: 'contract name?'
        },
        {
            type: 'checkbox',
            name: 'plugin',
            message: 'which plugins?',
            choices: [
                'client',
                'recoil',
                'react-query',
                'message-composer'
            ]
        }
    ];

    let { schema, out, name, plugin } = await prompt(questions, argv);
    if (!Array.isArray(plugin)) plugin = [plugin];


    ///////// REACT QUERY
    const questions2 = [];
    if (plugin.includes('react-query')) {
        [].push.apply(questions2, [
            {
                type: 'confirm',
                name: 'optionalClient',
                message: 'optionalClient?',
                default: false
            },
            {
                type: 'list',
                name: 'version',
                message: 'which react-query version?',
                default: 'v3',
                choices: ['v3', 'v4']
            }
        ])
    };
    const { optionalClient, version } = await prompt(questions2, argv);
    const questions3 = [];
    if (version === 'v4') {
        [].push.apply(questions3, [
            // currently we only support v4 for useMutation
            {
                type: 'confirm',
                name: 'mutations',
                message: 'Generate useMutation hooks?',
                default: false
            }

        ])
    };
    const { mutations } = await prompt(questions3, argv);
    ///////// END REACT QUERY


    const options: TSBuilderOptions = {
        types: {
            enabled: true
        },
        client: {
            enabled:
                plugin.includes('client') ||
                plugin.includes('recoil') ||
                plugin.includes('react-query')
        },
        reactQuery: {
            enabled: plugin.includes('react-query'),
            optionalClient,
            version,
            mutations
        },
        recoil: {
            enabled: plugin.includes('recoil'),
        },
        messageComposer: {
            enabled: plugin.includes('message-composer')
        }
    };

    await codegen({
        contracts: [
            {
                name,
                dir: schema
            }
        ],
        outPath: out,
        options
    })
};