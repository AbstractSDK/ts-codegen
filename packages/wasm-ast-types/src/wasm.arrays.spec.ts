import { importStmt } from './utils'
import generate from '@babel/generator';
import * as t from '@babel/types';

import message from './../../../__fixtures__/misc/schema/arrays.json';

import {
    createQueryClass,
    createQueryInterface,
    createExecuteClass,
    createExecuteInterface,
    createTypeInterface
} from './wasm'
import { RenderContext } from './utils/types';

const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}

const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}
const ctx = new RenderContext(message);

it('execute_msg_for__empty', () => {
    expectCode(createTypeInterface(
        ctx,
        message
    ))
})


it('query classes', () => {
    expectCode(createQueryClass(
        ctx,
        'SG721QueryClient',
        'SG721ReadOnlyInstance',
        message
    ))
});

it('execute classes array types', () => {
    expectCode(createExecuteClass(
        ctx,
        'SG721Client',
        'SG721Instance',
        null,
        message
    ))
});

it('execute interfaces no extends', () => {
    expectCode(createExecuteInterface(
        ctx,
        'SG721Instance',
        null,
        message
    ))
});
