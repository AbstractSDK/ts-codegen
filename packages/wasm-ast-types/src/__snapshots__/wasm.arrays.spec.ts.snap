// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute classes array types 1`] = `
"export class SG721Client implements SG721Instance {
  client: SigningCosmWasmClient;
  sender: string;
  contractAddress: string;

  constructor(client: SigningCosmWasmClient, sender: string, contractAddress: string) {
    this.client = client;
    this.sender = sender;
    this.contractAddress = contractAddress;
    this.updateEdges = this.updateEdges.bind(this);
  }

  updateEdges = async ({
    edges,
    nested,
    supernested
  }: {
    edges: number[][];
    nested: number[][][];
    supernested: string[][][][][][];
  }, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, funds?: readonly Coin[]): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      update_edges: {
        edges,
        nested,
        supernested
      }
    }, fee, memo, funds);
  };
}"
`;

exports[`execute interfaces no extends 1`] = `
"export interface SG721Instance {
  contractAddress: string;
  sender: string;
  updateEdges: ({
    edges,
    nested,
    supernested
  }: {
    edges: number[][];
    nested: number[][][];
    supernested: string[][][][][][];
  }, fee?: number | StdFee | \\"auto\\", memo?: string, funds?: readonly Coin[]) => Promise<ExecuteResult>;
}"
`;

exports[`execute_msg_for__empty 1`] = `"export type ExecuteMsg = ExecuteMsg;"`;

exports[`query classes 1`] = `
"export class SG721QueryClient implements SG721ReadOnlyInstance {
  client: CosmWasmClient;
  contractAddress: string;

  constructor(client: CosmWasmClient, contractAddress: string) {
    this.client = client;
    this.contractAddress = contractAddress;
    this.updateEdges = this.updateEdges.bind(this);
  }

  updateEdges = async ({
    edges,
    nested,
    supernested
  }: {
    edges: number[][];
    nested: number[][][];
    supernested: string[][][][][][];
  }): Promise<UpdateEdgesResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      update_edges: {
        edges,
        nested,
        supernested
      }
    });
  };
}"
`;
